randovania/interface_common/world_database.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+import asyncio
+import dataclasses
+import logging
+import uuid
+from pathlib import Path
+from typing import Iterable, Self
+
+from randovania.bitpacking.json_dataclass import JsonDataclass
+from randovania.interface_common.players_configuration import INVALID_UUID
+from randovania.lib import migration_lib, json_lib
+
+_MIGRATIONS = [
+    # lambda data: data,
+]
+CURRENT_VERSION = migration_lib.get_version(_MIGRATIONS)
+
+
+def migrate_to_current(data: dict):
+    return migration_lib.apply_migrations(data, _MIGRATIONS,
+                                          copy_before_migrating=True)
+
+
+@dataclasses.dataclass(frozen=True)
+class WorldServerData(JsonDataclass):
+    world_name: str
+    session_id: int
+    session_name: str
+
+
+def _combine_tuples(existing: tuple[int, ...], new_indices: Iterable[int]) -> tuple[int, ...]:
+    new = set(existing)
+    for it in new_indices:
+        new.add(it)
+    return tuple(sorted(new))
+
+
+@dataclasses.dataclass(frozen=True)
+class WorldData(JsonDataclass):
+    collected_locations: tuple[int, ...] = ()
+    uploaded_locations: tuple[int, ...] = ()
+    latest_message_displayed: int = 0
+    server_data: WorldServerData | None = None
+
+    def extend_collected_location(self, new_indices: Iterable[int]) -> Self:
+        return dataclasses.replace(
+            self,
+            collected_locations=_combine_tuples(self.collected_locations, new_indices),
+        )
+
+    def extend_uploaded_locations(self, new_indices: Iterable[int]) -> Self:
+        return dataclasses.replace(
+            self,
+            uploaded_locations=_combine_tuples(self.uploaded_locations, new_indices),
+        )
+
+
+class WorldDatabase:
+    _all_data: dict[uuid.UUID, WorldData]
+    _persist_path: Path
+
+    def __init__(self, persist_path: Path):
+        super().__init__()
+        self.logger = logging.getLogger(__name__)
+
+        persist_path.mkdir(parents=True, exist_ok=True)
+        self._persist_path = persist_path
+
+        self._remote_games = {}
```

Comment:
Is this field of `WorldDatabase` used somewhere? Can only find usages of the field with the same name of `MultiworldClient`

Reply:
oops